### PR TITLE
feat: support for more medium-priority array funcs

### PR DESCRIPTION
**Pull request by wangxye**

### CODE REVIEW

**Comment by ngjaying** [comment]
Thanks! The GitHub action seems broken now. We can rerun the checks later.

File: internal/binder/function/funcs_array.go [hunk 1]
```diff
@@ -105,4 +111,327 @@ func registerArrayFunc() {
...
+			res := make([]interface{}, 0, len(array))
+			if !ok {
+				return errorArrayFirstArgumentNotArrayError, false
+			}
+			for _, item := range array {
+				if item != args[1] {
+					res = append(res, item)
+				}
+			}
+			return res, true
```
Comment (Yisaer) [inline internal/binder/function/funcs_array.go:128]: I think we can avoid allocation on implementing `array_remove`

File: internal/binder/function/funcs_array.go [hunk 2]
```diff
@@ -105,4 +111,327 @@ func registerArrayFunc() {
...
+		fType: ast.FuncTypeScalar,
+		exec: func(ctx api.FunctionContext, args []interface{}) (interface{}, bool) {
+			array, ok := args[0].([]interface{})
+			if !ok {
+				return errorArrayFirstArgumentNotArrayError, false
+			}
+
+			for i := len(array) - 1; i >= 0; i-- {
+				if array[i] == args[1] {
+					return i + 1, true
```
Comment (Yisaer) [inline internal/binder/function/funcs_array.go:145]: we should keep 0-based index

File: internal/binder/function/funcs_array.go [hunk 3]
```diff
@@ -105,4 +111,327 @@ func registerArrayFunc() {
...
+		exec: func(ctx api.FunctionContext, args []interface{}) (interface{}, bool) {
+			array1, ok1 := args[0].([]interface{})
+			if !ok1 {
+				return errorArrayFirstArgumentNotArrayError, false
+			}
+			array2, ok2 := args[1].([]interface{})
+			if !ok2 {
+				return errorArraySecondArgumentNotArrayError, false
+			}
+			intersection := []interface{}{}
```
Comment (Yisaer) [inline internal/binder/function/funcs_array.go:194]: I think we can let max(len(array1),len(array2)) as the capacity of `intersection` here

File: internal/binder/function/funcs_array.go [hunk 4]
```diff
@@ -105,4 +111,327 @@ func registerArrayFunc() {
...
+		exec: func(ctx api.FunctionContext, args []interface{}) (interface{}, bool) {
+			array1, ok1 := args[0].([]interface{})
+			if !ok1 {
+				return errorArrayFirstArgumentNotArrayError, false
+			}
+			array2, ok2 := args[1].([]interface{})
+			if !ok2 {
+				return errorArraySecondArgumentNotArrayError, false
+			}
+			union := []interface{}{}
```
Comment (Yisaer) [inline internal/binder/function/funcs_array.go:226]: I think we can let len(array1) + len(array2) as the capacity of intersection here

File: internal/binder/function/funcs_array.go [hunk 5]
```diff
@@ -105,4 +111,327 @@ func registerArrayFunc() {
...
+				}
+			}
+
+			return union, true
+		},
+		val: func(ctx api.FunctionContext, args []ast.Expr) error {
+			return ValidateLen(2, len(args))
+		},
+	}
+	builtins["array_max"] = builtinFunc{
```
Comment (Yisaer) [inline internal/binder/function/funcs_array.go:248]: It seems `array_max` has lots of duplicated codes with `array_min`, could u remove the duplicated code and make it easy and clear?

File: internal/binder/function/funcs_array.go [hunk 6]
```diff
@@ -105,4 +111,327 @@ func registerArrayFunc() {
...
+		exec: func(ctx api.FunctionContext, args []interface{}) (interface{}, bool) {
+			array1, ok1 := args[0].([]interface{})
+			if !ok1 {
+				return errorArrayFirstArgumentNotArrayError, false
+			}
+			array2, ok2 := args[1].([]interface{})
+			if !ok2 {
+				return errorArraySecondArgumentNotArrayError, false
+			}
+			except := []interface{}{}
```
Comment (Yisaer) [inline internal/binder/function/funcs_array.go:349]: I think we can let len(array1) as the capacity of `except`

File: internal/binder/function/funcs_array.go [hunk 7]
```diff
@@ -105,4 +111,327 @@ func registerArrayFunc() {
...
+					step = -1
+				}
+			}
+
+			n := (stop-start)/step + 1
+
+			arr := make([]interface{}, n)
+			for i := range arr {
+				arr[i] = start + i*step
+			}
```
Comment (Yisaer) [inline internal/binder/function/funcs_array.go:424]: we should assert step can't be 0.

**Review by Yisaer — CHANGES_REQUESTED** [review]
rest LGTM

File: internal/binder/function/funcs_array.go (hunk 5 above)
Comment (wangxye) [reply internal/binder/function/funcs_array.go:248]: I ran into some function boundary handling details during the implementation, considering that the ARRAY_CREATE function allows elements of arrays to have different data types. Specifically, the following: Does the function also need to cover numbers in the form of strings, such as "123"? Is it possible to have an array of mixed strings, integers, and floats?

Comment (Yisaer) [reply internal/binder/function/funcs_array.go:259]: > Is it possible to have an array of mixed strings, integers, and floats?

I think it's possible.

**Review by Yisaer — APPROVED** [review]

**Review by superrxan — COMMENTED** [review]
please help update the docs for the new added functions
en_US: https://github.com/lf-edge/ekuiper/blob/master/docs/en_US/sqls/built-in_functions.md#array-functions
zh_CN: https://github.com/lf-edge/ekuiper/blob/master/docs/zh_CN/sqls/built-in_functions.md#%E5%88%97%E8%A1%A8%E5%87%BD%E6%95%B0

**Comment by wangxye** [comment]
> please help update the docs for the new added functions
> en_US: https://github.com/lf-edge/ekuiper/blob/master/docs/en_US/sqls/built-in_functions.md#array-functions
> zh_CN: https://github.com/lf-edge/ekuiper/blob/master/docs/zh_CN/sqls/built-in_functions.md#%E5%88%97%E8%A1%A8%E5%87%BD%E6%95%B0
sure, I will update it later.

**Comment by ngjaying** [comment]
> > please help update the docs for the new added functions
> > en_US: https://github.com/lf-edge/ekuiper/blob/master/docs/en_US/sqls/built-in_functions.md#array-functions
> > zh_CN: https://github.com/lf-edge/ekuiper/blob/master/docs/zh_CN/sqls/built-in_functions.md#%E5%88%97%E8%A1%A8%E5%87%BD%E6%95%B0
> 
> sure, I will update it later.

Will merge this one. Please open another PR for the doc update.

**Pull request merged**